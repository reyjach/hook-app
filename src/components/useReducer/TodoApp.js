import React, { useEffect, useReducer } from 'react';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {

    const value = JSON.parse(localStorage.getItem('todos'))

    if(value === null){
        return [];
    }
    return value;
    


    /*return [{
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    }]; */
}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init );

    

    useEffect(() => {
      localStorage.setItem('todos', JSON.stringify( todos ));
    }, [todos])
    

    const handleDelete = (todoId) => {

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = ( newTodo ) => {

        dispatch({
            type: 'add',
            payload:newTodo
        } );
    }

  return (
    <div>
        <h1>TodoApp ( { todos.length } )</h1>
        <hr />

        <div className="row">
            <div className="col-7">
                <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete}/>
            </div>

            <div className="col-5">
                <TodoAdd handleAddTodo={handleAddTodo}/>
            </div>

        </div>
    </div>
  )
}
