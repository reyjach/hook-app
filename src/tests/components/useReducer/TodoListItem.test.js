import { shallow } from "enzyme";
import { TodoListItem } from "../../../components/useReducer/TodoListItem";
import { demoTodo } from "../../fixture/demoTodo";

describe('Pruebas en el componente TodoListItem', () => { 

    

    const handleDelete = jest.fn();

    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoListItem 
            todo={ demoTodo[ 0 ] }
            i={ 0 }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />
    );

    beforeEach ( () => { 
        jest.clearAllMocks();

        const wrapper = shallow(
            <TodoListItem 
                todo={ demoTodo[ 0 ] }
                i={ 0 }
                handleDelete={ handleDelete }
                handleToggle={ handleToggle }
            />
        )
    })
    

    test('Debe de mostrarse correctamente', () => { 

        expect( wrapper ).toMatchSnapshot();

     });

     test('Debe de llamr a la funcion borrar', () => {

        wrapper.find('button').simulate('click');

        expect(handleDelete).toHaveBeenCalledWith(demoTodo[0].id);


     });

     test('Debe de llamar a la funcion toogle', () => { 

        wrapper.find('p').simulate('click');

        expect(handleToggle).toHaveBeenCalledWith(demoTodo[0].id);

      });

      test('Debe de mostrar el texto correcto', () => { 


        const p = wrapper.find('p');

        expect( p.text().trim() ).toBe(`${ 0 + 1 }. ${ demoTodo[0].desc }`);

       });

       test('Debe de tener la clase de Complete', () => { 

        const todo = demoTodo[0];

        todo.done = true;

        const wrapper = shallow(
            <TodoListItem 
                todo={ todo }
                i={0}
            />
        );

        expect(wrapper.find('p').hasClass('complete')).toBe(true);


        })
 })