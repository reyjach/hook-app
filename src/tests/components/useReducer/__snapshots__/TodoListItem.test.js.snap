// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente TodoListItem Debe de mostrarse correctamente 1`] = `
<Fragment>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      1
      . 
      Aprender react
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      borrar
    </button>
  </li>
</Fragment>
`;
